python/ray/tune/experiment/trial.py

```diff
@@ -749,23 +749,46 @@ def pickled_error_file(self):
             return None
         return os.path.join(self.local_path, self.run_metadata.pickled_error_filename)
 
-    def handle_error(self, exc: Optional[Union[TuneError, RayTaskError]] = None):
+    def _handle_restore_error(self, exc: _TuneRestoreError):
+        exc = exc.exc
+        if self.temporary_state.num_restore_failures >= int(
+            os.environ.get("TUNE_RESTORE_RETRY_NUM", 0)
+        ):
+            # Restore was unsuccessful, try again without checkpoint.
+            self.clear_checkpoint()
+            self.run_metadata.num_failures += 1
+        else:
+            self.temporary_state.num_restore_failures += 1
```

Comment:
This `handle_restore_error` happens when the call to `Trainable.restore` fails:
* This may be caused by a checkpoint download from cloud failing. Retrying without adding to the total failures counter may help here.
* There may be a bug in a user's `load_checkpoint` code. Retrying wouldn't help here.
* Function trainables don't do any logic in `restore`/`load_checkpoint`, leaving it to the user instead -- so this only really applies to class trainables.

Reply:
The default behavior is a little strange though: `TUNE_RESTORE_RETRY_NUM=0` --> failures during restore clear the checkpoint count toward `num_failures` and the run starts from scratch immediately.

If we remove this logic, the behavior becomes: failure during restore are treated normally and keep retrying until `max_failures`. I think it makes sense to remove this and `restoring_from` so that we have to keep track of less state in total. Let's do that in a separate PR.

Reply:
It seems the function here treats the restoration error differently as normal training error. I.e., there is a separate counter on restoration error that a consecutive of `TUNE_RESTORE_RETRY_NUM` restoration error will count as one `num_failures`.

I think it still makes sense if we keep this logic here. However, it doesn't makes much sense for us to clear or modify the latest checkpoint content. The `clear_checkpoint` function is mainly prepared for the cases that a corrupted checkpoint leading to restoration error. I agree it might be one reason of the problem, but not generally the only reason for a restoration failure. I think the easiest way for us to fix here is to remove the `clear_checkpoint` function . We can still keep the `handle_restore_error` function as a special case of all errors.

I.e., `TUNE_RESTORE_RETRY_NUM` restoration failures contributes to one `num_failures`. But we don't pre-assume or add special handling to fix the restoration error. It might be more likely due to a node preemption, that we don't need special handling, just by chance it may fail/ success. Add a few more retries can already help. We should not clear the latest checkpoint, which introduces extra complexity. In case of a corrupted latest checkpoint, we just let the job fail after `TUNE_RESTORE_RETRY_NUM` * `TUNE_RESTORE_NUM`

cc @justinvyu @matthewdeng, if it looks good, I can make a PR to fix this.